src/ram/src/ram.cpp

```diff
@@ -0,0 +1,609 @@
+            return false;
+          }
+          auto function = port->function();
+          return function->op() != sta::FuncExpr::op_not;
+        },
+        "tristate");
+  }
+
+  if (!and2_cell_) {
+    and2_cell_ = findMaster(
+        [this](sta::LibertyPort* port) {
+          if (!port->direction()->isOutput()) {
+            return false;
+          }
+          auto function = port->function();
+          return function && function->op() == sta::FuncExpr::op_and
+                 && function->left()->op() == sta::FuncExpr::op_port
+                 && function->right()->op() == sta::FuncExpr::op_port;
+        },
+        "and2");
+  }
+
+  if (!storage_cell_) {
+    // FIXME
+    storage_cell_ = findMaster(
+        [this](sta::LibertyPort* port) {
+          if (!port->direction()->isOutput()) {
+            return false;
+          }
+          auto function = port->function();
+          return function && function->op() == sta::FuncExpr::op_and
+                 && function->left()->op() == sta::FuncExpr::op_port
+                 && function->right()->op() == sta::FuncExpr::op_port;
+        },
+        "storage");
+  }
+
+  if (!clock_gate_cell_) {
+    clock_gate_cell_ = findMaster(
+        [this](sta::LibertyPort* port) {
+          return port->libertyCell()->isClockGate();
+        },
+        "clock gate");
+  }
+  // for input buffers
+  if (!buffer_cell_) {
+    buffer_cell_ = findMaster(
+        [this](sta::LibertyPort* port) {
+          return port->libertyCell()->isBuffer();
+        },
+        "buffer");
+  }
+}
+
+void RamGen::generate(const int bytes_per_word,
+                      const int word_count,
+                      const int read_ports,
+                      dbMaster* storage_cell,
+                      dbMaster* tristate_cell,
+                      dbMaster* inv_cell)
+{
+  const int bits_per_word = bytes_per_word * 8;
+  const std::string ram_name
+      = fmt::format("RAM{}x{}", word_count, bits_per_word);
+
+  logger_->info(RAM, 3, "Generating {}", ram_name);
+
+  storage_cell_ = storage_cell;
+  tristate_cell_ = tristate_cell;
+  inv_cell_ = inv_cell;
+  and2_cell_ = nullptr;
+  clock_gate_cell_ = nullptr;
+  buffer_cell_ = nullptr;
+  findMasters();
+
+  auto chip = db_->getChip();
+  if (!chip) {
+    chip = odb::dbChip::create(db_);
+  }
+
+  block_ = chip->getBlock();
+  if (!block_) {
+    block_ = odb::dbBlock::create(chip, ram_name.c_str());
+  }
+
+  // 9 tracks for 8 bits per word plus
+  // cell for WE AND gate/inverter
+  // extra 1 is for decoder cells
+  int col_cell_count = bytes_per_word * 9;
+  Grid ram_grid(odb::horizontal, col_cell_count + 1);
+
+  auto clock = makeBTerm("clk");
+
+  vector<dbBTerm*> write_enable(bytes_per_word, nullptr);
+  for (int byte = 0; byte < bytes_per_word; ++byte) {
+    auto in_name = fmt::format("we[{}]", byte);
+    write_enable[byte] = makeBTerm(in_name);
+  }
+
+  // input bterms
+  int numImputs = std::log2(word_count);
```

Comment:
`std::log2` returns a float, which is then truncated when assigning to an `int`. e.g. `std::log2(5)` will return ~2.3, which gets truncated to 2. You need to round up the number of bits required.
```suggestion
  int numImputs = std::ceil(std::log2(word_count));
```